### PR TITLE
feat: controller budget management

### PR DESCRIPTION
**Pull request by gabriela-garcia-jalafund**
## Trello Ticket

[ID-03](https://trello.com/c/kzOwelSe/38-id-03-crud-for-budget-management)

## What did I do?

 I implemented the budget management feature for our application. The main changes include:

- Created API Endpoints: Developed endpoints for creating, retrieving, updating, and deleting budgets.
- Budget Model: Introduced a new Budget model to represent budget data in the database.

## Why did I do this?

These changes were necessary to provide users with the ability to manage their budgets effectively. Previously, the application lacked budget management functionality, which limited users' ability to track and plan their finances. By implementing these features, we solve the problem of budget tracking, adding significant value to the project by enhancing user engagement and satisfaction.

## How did I do it?

To implement these changes, I followed these steps:

- API Development: Used Express.js to create RESTful API endpoints for budget management (POST, GET, PUT, DELETE).
- Database Integration: Utilized MongoDB to store budget data, defining a schema for the Budget model.

### CODE REVIEW

File: src/ExpenseTrackerGrupo2/DataAccess/Repositories/Concretes/BudgetRepository.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+public class BudgetRepository : BaseRepository<Budget>, IBudgetRepository
+{
+    public BudgetRepository(IDbConnectionFactory dbConnectionFactory) : base(dbConnectionFactory) { }
+
+       public async Task<bool> SetBudget(Budget budget)
+        {
+            using var connection = await _dbConnectionFactory.CreateConnectionAsync();
+            var query = "INSERT INTO budget (budget_id, user_id, month, year, amount) VALUES (@BudgetId, @UserId, @Month, @Year, @Amount)";
+            var rowsAffected = await connection.ExecuteAsync(query, budget);
+            return rowsAffected > 0; // Return true if the insert was successful
```
Comment (salet-gutierrez-jalafund) [inline src/ExpenseTrackerGrupo2/DataAccess/Repositories/Concretes/BudgetRepository.cs]: Delete message

**Review by salet-gutierrez-jalafund — CHANGES_REQUESTED** [review]

File: src/ExpenseTrackerGrupo2/Controllers/BudgetController.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
+using Microsoft.AspNetCore.Mvc;
+using ExpenseTrackerGrupo2.Business.Services.Interfaces;
+using ExpenseTrackerGrupo2.Business.Services.Mappers.Requests.Budgets;
+
+namespace ExpenseTrackerGrupo2.API.Controllers
+{
```
Comment (salet-gutierrez-jalafund) [inline src/ExpenseTrackerGrupo2/Controllers/BudgetController.cs:10]: enter a ; instead of {}

**Review by salet-gutierrez-jalafund — CHANGES_REQUESTED** [review]

File: src/ExpenseTrackerGrupo2/DataAccess/Repositories/Concretes/BudgetRepository.cs (hunk 1 above)
Comment (gabriela-garcia-jalafund) [reply src/ExpenseTrackerGrupo2/DataAccess/Repositories/Concretes/BudgetRepository.cs]: Done

File: src/ExpenseTrackerGrupo2/DataAccess/Repositories/Interfaces/IBudgetRepository.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+using ExpenseTrackerGrupo2.DataAccess.Entities;
+
+namespace ExpenseTrackerGrupo2.DataAccess.Concretes
+{
```
Comment (salet-gutierrez-jalafund) [inline src/ExpenseTrackerGrupo2/DataAccess/Repositories/Interfaces/IBudgetRepository.cs:6]: enter a ; instead of {}

**Review by salet-gutierrez-jalafund — CHANGES_REQUESTED** [review]

File: src/ExpenseTrackerGrupo2/Controllers/BudgetController.cs (hunk 2 above)
Comment (gabriela-garcia-jalafund) [reply src/ExpenseTrackerGrupo2/Controllers/BudgetController.cs:10]: Done

File: src/ExpenseTrackerGrupo2/DataAccess/Repositories/Concretes/BudgetRepository.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            var rowsAffected = await connection.ExecuteAsync(query, budget);
+            return rowsAffected > 0;
+        }
+
+        public async Task<Budget> GetBudgetByUserIdAndMonth(Guid userId, int month, int year)
+        {
+            using var connection = await _dbConnectionFactory.CreateConnectionAsync();
+            var query = "SELECT * FROM budget WHERE user_id = @UserId AND month = @Month AND year = @Year";
+            var budget = await connection.QuerySingleOrDefaultAsync<Budget>(query, new { UserId = userId, Month = month, Year = year });
+            return budget;
```
Comment (alex-paca-jalafund) [inline src/ExpenseTrackerGrupo2/DataAccess/Repositories/Concretes/BudgetRepository.cs]: Fix warnings pls for possible null reference on return

**Review by alex-paca-jalafund — CHANGES_REQUESTED** [review]

File: src/ExpenseTrackerGrupo2/DataAccess/Repositories/Concretes/BudgetRepository.cs (hunk 4 above)
Comment (gabriela-garcia-jalafund) [reply src/ExpenseTrackerGrupo2/DataAccess/Repositories/Concretes/BudgetRepository.cs]: Done

File: src/ExpenseTrackerGrupo2/Controllers/BudgetController.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
+using Microsoft.AspNetCore.Mvc;
+using ExpenseTrackerGrupo2.Business.Services.Interfaces;
+using ExpenseTrackerGrupo2.Business.Services.Mappers.Requests.Budgets;
+
+namespace ExpenseTrackerGrupo2.API.Controllers;
+
+[ApiController]
+[Route("api/v1/budget/[controller]")]
```
Comment (salet-gutierrez-jalafund) [inline src/ExpenseTrackerGrupo2/Controllers/BudgetController.cs]: image

that it only contains either budget or Budget but that it does not contain both lowercase or uppercase letters, it is better if you put the uppercase letter.

**Review by salet-gutierrez-jalafund — CHANGES_REQUESTED** [review]

File: src/ExpenseTrackerGrupo2/DataAccess/Repositories/Interfaces/IBudgetRepository.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+using ExpenseTrackerGrupo2.DataAccess.Entities;
+
+namespace ExpenseTrackerGrupo2.DataAccess.Concretes;
+
+    public interface IBudgetRepository : IBaseRepository<Budget>
```
Comment (salet-gutierrez-jalafund) [inline src/ExpenseTrackerGrupo2/DataAccess/Repositories/Interfaces/IBudgetRepository.cs]: that does not have space, you have to be at the same height as the namespace

**Review by salet-gutierrez-jalafund — CHANGES_REQUESTED** [review]

File: src/ExpenseTrackerGrupo2/Controllers/BudgetController.cs (hunk 5 above)
Comment (gabriela-garcia-jalafund) [reply src/ExpenseTrackerGrupo2/Controllers/BudgetController.cs]: Done

File: src/ExpenseTrackerGrupo2/DataAccess/Repositories/Interfaces/IBudgetRepository.cs (hunk 6 above)
Comment (gabriela-garcia-jalafund) [reply src/ExpenseTrackerGrupo2/DataAccess/Repositories/Interfaces/IBudgetRepository.cs]: Done

**Comment by salet-gutierrez-jalafund** [comment]
[ID-03: CRUD for Budget Management](https://trello.com/c/kzOwelSe/38-id-03-crud-for-budget-management)

**Review by alex-paca-jalafund — CHANGES_REQUESTED** [review]
{
  "budgetAmount": 20000000,
  "month": "2024-09",
  "user_id": "550e8400-e29b-41d4-a716-446655440000"
}

Method POST its not working:

{
  "type": "https://tools.ietf.org/html/rfc9110#section-15.5.1",
  "title": "One or more validation errors occurred.",
  "status": 400,
  "errors": {
    "request": [
      "The request field is required."
    ],
    "$.month": [
      "The JSON value could not be converted to ExpenseTrackerGrupo2.Business.Services.Mappers.Requests.Budgets.BudgetCreateRequest. Path: $.month | LineNumber: 2 | BytePositionInLine: 20."
    ]
  },
  "traceId": "00-e664ebe4c4d023a7370b361837f95438-544a8e5cff3021a1-00"

**Review by salet-gutierrez-jalafund — APPROVED** [review]
good job

**Review by alex-paca-jalafund — APPROVED** [review]
All the crud operations are working correctly.

**Pull request merged**